File: Praticas/3/entropia.py
import sys
import math
from collections import Counter

def calcular_entropia(caminho_ficheiro):
    try:
        # Lê o ficheiro em formato binário ('rb') para ler os bytes reais
        with open(caminho_ficheiro, 'rb') as f:
            dados = f.read()
    except FileNotFoundError:
        print(f"Erro: O ficheiro '{caminho_ficheiro}' não foi encontrado.")
        return

    total_bytes = len(dados)
    if total_bytes == 0:
        print("O ficheiro está vazio. Entropia: 0")
        return

    # Conta a frequência de cada byte (0 a 255)
    contagem_bytes = Counter(dados)

    # Calcula a entropia usando a fórmula de Shannon (base 10)
    entropia = 0.0
    for contagem in contagem_bytes.values():
        probabilidade = contagem / total_bytes
        entropia -= probabilidade * math.log10(probabilidade)

    # Calcula a entropia máxima para 256 valores possíveis (n = 256)
    entropia_maxima = math.log10(256)

    print(f"--- Resultados para '{caminho_ficheiro}' ---")
    print(f"Tamanho do ficheiro: {total_bytes} bytes")
    print(f"Entropia calculada:  {entropia:.4f}")
    print(f"Entropia máxima:     {entropia_maxima:.4f}")

if __name__ == "__main__":
    # Verifica se o utilizador passou o nome do ficheiro como argumento
    if len(sys.argv) != 2:
        print("Uso correto: python entropia.py <nome_do_ficheiro>")
    else:
        calcular_entropia(sys.argv[1])

File: Praticas/3/entropia.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void calcular_entropia(const char *caminho_ficheiro) {
    // Abre o ficheiro em modo binário de leitura ("rb")
    FILE *ficheiro = fopen(caminho_ficheiro, "rb");
    if (ficheiro == NULL) {
        printf("Erro: O ficheiro '%s' não foi encontrado ou não pode ser lido.\n", caminho_ficheiro);
        return;
    }

    // Array para contar a frequência de cada byte (0 a 255)
    long contagem_bytes[256] = {0};
    long total_bytes = 0;
    int byte;

    // Lê o ficheiro byte a byte até chegar ao fim (EOF)
    while ((byte = fgetc(ficheiro)) != EOF) {
        contagem_bytes[byte]++;
        total_bytes++;
    }

    fclose(ficheiro);

    // Verifica se o ficheiro está vazio
    if (total_bytes == 0) {
        printf("O ficheiro está vazio. Entropia: 0\n");
        return;
    }

    // Calcula a entropia usando a fórmula de Shannon (base 10)
    double entropia = 0.0;
    for (int i = 0; i < 256; i++) {
        if (contagem_bytes[i] > 0) {
            double probabilidade = (double)contagem_bytes[i] / total_bytes;
            entropia -= probabilidade * log10(probabilidade);
        }
    }

    // Calcula a entropia máxima para 256 valores possíveis (n = 256)
    double entropia_maxima = log10(256.0);

    // Imprime os resultados
    printf("--- Resultados para '%s' ---\n", caminho_ficheiro);
    printf("Tamanho do ficheiro: %ld bytes\n", total_bytes);
    printf("Entropia calculada:  %.4f\n", entropia);
    printf("Entropia máxima:     %.4f\n", entropia_maxima);
}

int main(int argc, char *argv[]) {
    // Verifica se o utilizador passou o nome do ficheiro como argumento
    if (argc != 2) {
        printf("Uso correto: ./entropia <nome_do_ficheiro>\n");
        return 1;
    }

    calcular_entropia(argv[1]);
    return 0;
}